### PR TITLE
PWN-6443 - Off-ramp. API for getting deposit wallet address

### PR DESCRIPTION
**Pull request by gslevinkov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-6443

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt [hunk 1]
```diff
@@ -60,18 +64,36 @@ class MoonpaySellRemoteRepository(
...
-            response = moonpayServerSideApi.getUserSellTransactions(userAddress.base58Value),
-            transactionOwnerAddress = userAddress
+            response = response,
+            depositWallets = depositWallets,
+            selectedFiat = getSellFiatCurrency(),
+            transactionOwnerAddress = userAddress,
         )
     }
 
+    private suspend fun getDepositWalletsForTransactions(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt:79]: for each WaitingForDeposit transaction we fetch deposit wallet via another request

File: app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRepositoryErrorMapper.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package org.p2p.wallet.moonpay.repository.sell
+
+import okio.IOException
+import org.p2p.wallet.infrastructure.network.data.ServerException
+import org.p2p.wallet.infrastructure.network.moonpay.MoonpayErrorResponseType
+import org.p2p.wallet.moonpay.model.MoonpaySellError
+
+class MoonpaySellRepositoryErrorMapper {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRepositoryErrorMapper.kt:8]: just extracted to a separate class to remove lines in the main mapper

File: app/src/main/java/org/p2p/wallet/sell/interactor/SellTransactionViewDetailsMapper.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+import android.content.res.Resources
+import org.p2p.core.utils.formatToken
+import org.p2p.core.utils.formatUsd
+import org.p2p.wallet.R
+import org.p2p.wallet.moonpay.model.SellTransaction
+import org.p2p.wallet.sell.ui.lock.SellTransactionViewDetails
+
+class SellTransactionViewDetailsMapper(private val resources: Resources) {
+    fun fromDomain(transactions: List<SellTransaction>): List<SellTransactionViewDetails> {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/interactor/SellTransactionViewDetailsMapper.kt:11]: reusable logic for mapping for HistoryPresenter and TokenHistoryPresenter, both need SellTransactionViewDetails

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt [hunk 4]
```diff
@@ -77,19 +76,22 @@ class SellPayloadPresenter(
         if (userTransactionInProcess != null) {
             // make readable in https://p2pvalidator.atlassian.net/browse/PWN-6354
             val amounts = userTransactionInProcess.amounts
-            val details = SellTransactionDetails(
-                status = userTransactionInProcess.status,
-                formattedSolAmount = amounts.tokenAmount.formatToken(),
-                formattedUsdAmount = userTransactionInProcess.amounts.getAmountFromFiat(currentFiat).formatUsd(),
-                receiverAddress = WRAPPED_SOL_MINT // todo replace with correct address
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt:84]: replaced :)))

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt [hunk 5]
```diff
@@ -109,7 +111,7 @@ class HistoryPresenter(
     }
 
     private suspend fun fetchMoonpayTransactions() {
-        val moonpayItems = sellInteractor.loadUserSellTransactionsDetails()
+        val moonpayItems = sellTransactionsMapper.fromDomain(sellInteractor.loadUserSellTransactions())
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:114]: Why are you doing mapping in presenter?

Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:114]: If you need items for adapter, maybe you'll move mapping to adapter?

File: app/src/main/java/org/p2p/wallet/infrastructure/network/interceptor/MoonpayRequestException.kt [hunk 6]
```diff
@@ -1,13 +1,14 @@
...
+import okio.IOException
 import org.p2p.wallet.infrastructure.network.moonpay.MoonpayErrorResponseType
 import javax.net.ssl.HttpsURLConnection
 
 class MoonpayRequestException(
     val httpCode: Int,
     val errorType: MoonpayErrorResponseType,
     override val message: String
-) : Throwable(message) {
+) : IOException(message) {
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/interceptor/MoonpayRequestException.kt:11]: Why?

File: app/src/main/java/org/p2p/wallet/moonpay/model/SellTransaction.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        val createdAt: String,
+        val updatedAt: String,
+        val accountId: String,
+        val customerId: String,
+        val bankAccountId: String,
+        val externalTransactionId: String?,
+        val externalCustomerId: String?,
+        val countryAbbreviation: String,
+        val stateAbbreviation: String?
+    )
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/moonpay/model/SellTransaction.kt:21]: please move it to a separate file

File: app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt [hunk 8]
```diff
@@ -60,18 +64,36 @@ class MoonpaySellRemoteRepository(
...
+        transactions: List<MoonpaySellTransactionShortResponse>
+    ): List<MoonpaySellTransactionDepositWalletResponse> =
+        transactions.filter { it.status == SellTransactionStatus.WAITING_FOR_DEPOSIT }
+            .mapNotNull {
+                kotlin.runCatching {
+                    moonpayClientSideApi.getSellTransactionDepositWalletById(
+                        transactionId = it.transactionId,
+                        apiKey = BuildConfig.moonpayKey
+                    )
+                }.getOrNull() // skip if couldn't fetch
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt:89]: imho, we need to change our code style rules about first operation. 
```
transactions
     .filter()
     .mapNotNull()
     .getOrNull()
```

is better

Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt:89]: Are you sure you need `runCatching` here?

File: app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt [hunk 9]
```diff
@@ -60,18 +64,36 @@ class MoonpaySellRemoteRepository(
...
+            response = response,
+            depositWallets = depositWallets,
+            selectedFiat = getSellFiatCurrency(),
+            transactionOwnerAddress = userAddress,
         )
     }
 
+    private suspend fun getDepositWalletsForTransactions(
+        transactions: List<MoonpaySellTransactionShortResponse>
+    ): List<MoonpaySellTransactionDepositWalletResponse> =
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt:81]: The repository should return domain models

**Review by 1sakov — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt (hunk 5 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:114]: yeah, good idea
i think we don't have a strict rule where to do mapping for the adapter, i love the idea to do it in adapter

File: app/src/main/java/org/p2p/wallet/infrastructure/network/interceptor/MoonpayRequestException.kt (hunk 6 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/infrastructure/network/interceptor/MoonpayRequestException.kt:11]: Strangely, Okhttp doesn't accept correctly if you throw `Throwable`, it works okay if the exception is `IOException`
I read on the internet that any error / exception in interceptor that is not IOException - means something FATAL for OkHttp and it crashes the app.

File: app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt (hunk 8 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt:89]: i need it, but im not sure how to do it properly, i'll come to dm

File: app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt (hunk 9 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt:81]: it's a private method :)

File: app/src/main/java/org/p2p/wallet/history/HistoryModule.kt [hunk 10]
```diff
@@ -26,88 +26,31 @@ import org.p2p.wallet.history.ui.token.TokenHistoryContract
...
-                receiveAnalytics = get(),
-                swapAnalytics = get(),
-                analyticsInteractor = get(),
-                environmentManager = get(),
-                sendAnalytics = get(),
-                sellInteractor = get()
-            )
-        } bind HistoryContract.Presenter::class
+        single { get<Retrofit>(named(RpcModule.RPC_RETROFIT_QUALIFIER)).create(RpcHistoryApi::class.java) }
+        singleOf(::TransactionDetailsRpcRepository) bind TransactionDetailsRemoteRepository::class
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/HistoryModule.kt:54]: maybe you'd better to create an api within `single` block of repository? 
You would avoid adding one more dependency in the graph

**Review by 1sakov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/history/HistoryModule.kt (hunk 10 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/history/HistoryModule.kt:54]: well, okay, will fix it in another pr

**Review by kabramovp2p — APPROVED** [review]

**Pull request merged**